pkg/northbound/gnmi/subscribe.go

```diff
@@ -44,79 +52,105 @@ func (s *Server) Subscribe(stream gnmi.GNMI_SubscribeServer) error {
 	changesChan, err := mgr.Dispatcher.RegisterNbi(hash)
 	if err != nil {
 		log.Println("Subscription present: ", err)
-		return err
+		return status.Error(codes.Internal, err.Error())
 	}
 	opStateChan, err := mgr.Dispatcher.RegisterOpState(hash)
 	if err != nil {
 		log.Println("Subscription present: ", err)
-		return err
+		return status.Error(codes.Internal, err.Error())
 	}
+	resChan := make(chan result)
 	//Handles each subscribe request coming into the server, blocks until a new request or an error comes in
-	for {
-		in, err := stream.Recv()
-		if err == io.EOF {
-			log.Println("Subscription Terminated")
-			//Ignoring Errors during removal
-			mgr.Dispatcher.UnregisterNbi(hash)
-			mgr.Dispatcher.UnregisterOperationalState(hash)
-			return nil
-		}
+	go func() {
+		for {
+			in, err := stream.Recv()
+			if err == io.EOF {
+				log.Println("Subscription Terminated")
+				//Ignoring Errors during removal
+				mgr.Dispatcher.UnregisterNbi(hash)
+				mgr.Dispatcher.UnregisterOperationalState(hash)
+				resChan <- result{success: true, err: nil}
+				break
+			}
 
-		if err != nil {
-			log.Println("Error in subscription", err)
-			//Ignoring Errors during removal
-			mgr.Dispatcher.UnregisterNbi(hash)
-			mgr.Dispatcher.UnregisterOperationalState(hash)
-			return err
-		}
+			if err != nil {
+				code, ok := status.FromError(err)
+				if ok && code.Code() == codes.Canceled {
+					log.Println("Subscription Terminated")
+					resChan <- result{success: true, err: nil}
```

Comment:
You don't exit here?

Reply:
no, the the result gets issued and the for loop picks it up. I break the for effectively ending and enabling garbage collection on the routine.